patches/api/0421-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,2335 @@
++     * @param source         The entity source or null
++     * @param sourcePosition The source position
++     *
++     * @return A builder to then build a new {@link DamageOrigin}
++     */
++    @NotNull
++    static DamageOriginBuilder of(@NotNull DamageType type, @Nullable Entity directSource, @Nullable Entity source, @Nullable Position sourcePosition) {
++        return Bukkit.createDamageSource(type, directSource, source, sourcePosition);
++    }
++
++    /**
++     * Creates a generic {@link DamageOriginBuilder} with optionally
++     * a direct entity, an entity and a position.
++     * For example, in the case of a projectile dealing damage,
++     * the direct source will be the projectile and the source will be
++     * the thrower.
++     *
++     * @param type         The damage type
++     * @param directSource The entity direct source, the nearest entity
++     *                     that will damage the target or null
++     * @param source       The entity source or null
++     *
++     * @return A builder to then build a new {@link DamageOrigin}
++     */
++    @NotNull
++    static DamageOriginBuilder of(@NotNull DamageType type, @Nullable Entity directSource, @Nullable Entity source) {
++        return of(type, directSource, source, null);
++    }
++
++    /**
++     * Creates a generic {@link DamageOriginBuilder} with optionally
++     * a position.
++     * Source position is used for explosion related damage.
++     *
++     * @param type           The damage type
++     * @param sourcePosition The source position
++     *
++     * @return A builder to then build a new {@link DamageOrigin}
++     */
++    @NotNull
++    static DamageOriginBuilder of(@NotNull DamageType type, @Nullable Position sourcePosition) {
++        return of(type, null, null, sourcePosition);
++    }
++
++    /**
++     * Creates a generic {@link DamageOriginBuilder}.
++     *
++     * @param type The damage type
++     *
++     * @return A builder to then build a new {@link DamageOrigin}
++     */
++    @NotNull
++    static DamageOriginBuilder of(@NotNull DamageType type) {
++        return of(type, null, null, null);
++    }
++
++    /**
++     * Gets the type of this damage origin.
++     * The type hold a lot of generic information like
++     * the hurt sound, death message format and more...
++     *
++     * @return the damage type
++     */
++    @NotNull
++    DamageType getType();
++
++    /**
++     * Checks if this damage origin is considered as sweep damage.
++     * A sweep damage generally hits multiple targets in a radius
++     * close to the attacker
++     *
++     * @return true for sweep damage
++     */
++    boolean isSweep();
++
++    /**
++     * Checks if this damage origin will do a critical hit or not.
++     * A critical hit will generally deal more damage to the target
++     *
++     * @return True for a critical damage
++     */
++    boolean isCritical();
```

Comment:
These two are logic implemented by the api in order to pass context. I feel like these should not be included, and should instead be passed through the event context.

Reply:
Do you mean having the event context inside of the damage origin for plugin one or just a new option for damage method without saving them in the damage origin (just one the underlying damage source) ?

Reply:
I am not sure. But take a look at how these are used in the api. It's simply used for event context.